### PR TITLE
add custom_jvp for jax.nn.softmax

### PR DESCRIPTION
**Pull request by mattjj**
This approach avoids saving the jnp.exp(...) value.

This implementation is basically the same as [the one in TF](https://github.com/tensorflow/tensorflow/blob/1df4022ca1ad8b131765148e2340bdea9ffdaec2/tensorflow/python/ops/nn_grad.py#L282-L304).

Thanks to @blakehechtman for pointing this out and advising.

We expect this change to be an improvement for all users, since it can improve stability and may use less memory (though in many cases XLA optimizations mean the memory usage may not really change). However, since it's changing a heavily relied-upon function, we're including a flag which can be set to restore the old behavior. We'll keep the flag around until we're sure this change doesn't cause regressions.

To restore the old behavior for now, we can:
1. use the context manager `with jax.softmax_custom_jvp(False): ...`, or
2. use the shell environment variable `JAX_SOFTMAX_CUSTOM_JVP=0`, or
3. set the config option `jax.config.update('jax_softmax_custom_jvp', False)`, or
4. if parsing flags with absl, use the flag `--jax_softmax_custom_jvp=False`.

The config option is thread- and jit-cache-safe, as usual.

Before:
image

After:
image

TODO:
* [x] add a flag to switch this off (doesn't have to be a nice or documented flag, but just "break glass in case of emergency" while we wait for things to shake out)
* [x] show that the saved residuals are different (in a test, or just in this PR message)
* [x] adapt any google-internal goldens that this might affect

### CODE REVIEW

File: jax/_src/nn/functions.py [hunk 1]
```diff
@@ -344,9 +345,15 @@ def softmax(x: Array,
       when :code:`where` is not None.
   """
   x_max = jnp.max(x, axis, where=where, initial=initial, keepdims=True)
-  unnormalized = jnp.exp(x - lax.stop_gradient(x_max))
```
Comment (mattjj) [inline jax/_src/nn/functions.py:347]: We no longer need `lax.stop_gradient` here because we no longer differentiate through this implementation, instead relying on the `custom_jvp` rule defined below.

**Review by blakehechtman — APPROVED** [review]

**Review by jekbradbury — APPROVED** [review]

**Comment by NeilGirdhar** [comment]
@mattjj  Are you guys planning on making this default to true?
